python/KAPEL.py

```diff
@@ -159,43 +188,14 @@ def rearrange(x):
         # this produces each of the (key, value) tuples in the list
         yield item['metric']['pod'], float(item['value'][1])
 
-# process a time period (do prom query, process data, write output)
-# takes a KAPELConfig object and one element of output from get_time_periods
-# Remember Prometheus queries go backwards: the time instant is the end, go backwards from there.
-def process_period(config, period):
-    period_start = period['instant'] + dateutil.relativedelta.relativedelta(seconds=-period['range_sec'])
-    print(
-        f"Processing year {period['year']}, month {period['month']}, "
-        f"querying from {period['instant'].isoformat()} and going back {period['range_sec']} s to {period_start.isoformat()}."
-    )
-    queries = QueryLogic(queryRange=(str(period['range_sec']) + 's'), namespace=config.namespace)
-
-    # SSL generally not used for Prometheus access within a cluster
-    # Docs on instant query API: https://prometheus.io/docs/prometheus/latest/querying/api/#instant-queries
-    prom = PrometheusConnect(url=config.prometheus_server, disable_ssl=True)
-    prom_connect_params = {'time': period['instant'].isoformat(), 'timeout': config.query_timeout}
-
-    raw_results, results, result_lengths = {}, {}, []
-    # iterate over each query (cputime, starttime, endtime, cores) producing raw_results['cputime'] etc.
-    for query_name, query_string in vars(queries).items():
-        # Each of these raw_results is a list of dicts. Each dict in the list represents an individual data point, and contains:
-        # 'metric': a dict of one or more key-value pairs of labels, one of which is the pod name.
-        # 'value': a list in which the 0th element is the timestamp of the value, and 1th element is the actual value we're interested in.
-        print(f'Executing {query_name} query: {query_string}')
-        t1 = timer()
-        raw_results[query_name] = prom.custom_query(query=query_string, params=prom_connect_params)
-        t2 = timer()
-        results[query_name] = dict(rearrange(raw_results[query_name]))
-        result_lengths.append(len(results[query_name]))
-        t3 = timer()
-        print(f'Query finished in {t2 - t1} s, processed in {t3 - t2} s. Got {len(results[query_name])} items from {len(raw_results[query_name])} results. Peak RAM usage: {resource.getrusage(resource.RUSAGE_SELF).ru_maxrss}K.')
-        del raw_results[query_name]
 
+def record_summarized_period(config, period_start, year, month, results):
+    """ Record the sum of usage across all pods in the given time period. """
     cputime = results['cputime']
     endtime = results['endtime']
     starttime = results['starttime']
     cores = results['cores']
-
+    result_lengths = max(len(l) for l in results.values())
```

Comment:
@mwestphall Previously result_lengths was an array containing the number of results of the cputime, starttime, endtime, cores queries (and now also cpuusage and memory).
So shouldn't result_lengths be perhaps something like this?
`result_lengths = [(len(l) for l in results.values())]`   (I didn't test the syntax.)

Otherwise the assertion on line 203 might fail because first it would take the max of the array but then it is checked against the min of the array.

Reply:
@rptaylor good catch here, the `min(max([<array>]))` that's effectively being done here gives a TypeError. The change you suggested should work.